feat: Prac0 Intro to ML using JAX

KaleabTessera opened the pull request:
Draft PR for Prac 0, [colab link](https://colab.research.google.com/github/deep-learning-indaba/indaba-pracs-2022/blob/prac1-intro-ml-using-jax/0_intro_to_ml_using_jax.ipynb)
- Part 1 is done. Please run through this in colab and let me know how long it takes and if you have any comments and suggestion. 
- Part 2 is also done. Added a Haiku and Optax section.

JavierAntoran commented:
I did the intro prac part 1.

Timing:

 It took me 42 minutes, although I got distracted reading through documentation and with slack messages various times  (Perhaps this simulates the lively in-person indaba environment.) Without stopping, I think it would have taken 30 minutes. If I had not read all the cells and understood all of the code (since I could guess what it was going to do from experience with jax / numpy) I think it would have taken 15 minutes.


Comments / potential improvements:

 - I would uncollapse cells by default because otherwise it is possible to get confused and skip a section

 - In the random key explanation for jax, it is not completely obvious that in `key, subkey = random.split(key)` , the returned key is different from the inputted key (since both have the same name). This could cause some confusion.
 
 - In the ReLU example, I would make the provided solution use` jnp.where`  since the `jnp.maximum`  solution is trivial. This way people will probably use `jnp.maximum` in their solution but can learn something from the solved exercise as well.
 
 - The `min_max(batched_x)` example has the solution already coded up and inputed by default. This is a bug I think.
 
 - For the parallelisation of the `min_max(batched_x)` section, the matrices used are huge. I let this run for 7 minutes on CPU and stopped it. I would make the leading dimension 500 instead of 5000 since otherwise everyone will need GPU allocated. I could not get GPU or TPU today for some reason. There are 2 or 3 cells where this is a problem.
 
 - The quiz form would not load in the notebook for me and I had to click the link to go to the quiz page explicitly

JamesAllingham commented on 0_intro_to_jax_for_ml.ipynb:
```diff
@@ -0,0 +1,2081 @@
+{
```
<p>Nit: <code>Grad</code> -&gt; <code>grad</code>, <code>JIT</code> -&gt; <code>jit</code> .    </p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/deep-learning-indaba/indaba-pracs-2022/pull/4/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='0'/>

KaleabTessera replied:
<p>Updated. Thanks! </p>

JamesAllingham commented on 0_intro_to_jax_for_ml.ipynb:
```diff
@@ -0,0 +1,2081 @@
+{
```
<p>Are you intentionally having this (and the next) cell opened by default?</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/deep-learning-indaba/indaba-pracs-2022/pull/4/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='2'/>

KaleabTessera replied:
<p>Made it hidden now. Thanks!</p>

JamesAllingham commented on 0_intro_to_jax_for_ml.ipynb:
```diff
@@ -0,0 +1,2081 @@
+{
```
<p>This cell was open by default for me -- probably don't want that!</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/deep-learning-indaba/indaba-pracs-2022/pull/4/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='14'/>

KaleabTessera replied:
<p>Fixed. </p>

JamesAllingham commented on 0_intro_to_jax_for_ml.ipynb:
```diff
@@ -0,0 +1,2081 @@
+{
```
<pre class="ql-syntax" spellcheck="false">Line #1.    # @title Answer to code task (Try not to peek until you've given it a good try!') 
</pre><p>Nit: extra "'" at the end.</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/deep-learning-indaba/indaba-pracs-2022/pull/4/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='14'/>

KaleabTessera replied:
<p>Done. </p>

JamesAllingham commented on 0_intro_to_jax_for_ml.ipynb:
```diff
@@ -0,0 +1,2081 @@
+{
```
<p>Nit: For the second bullet point I'd say "Jax handles <strong>randomness</strong> <strong>explicitly</strong>.", for clarity.</p><p> </p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/deep-learning-indaba/indaba-pracs-2022/pull/4/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='15'/>

KaleabTessera replied:
<p>Makes sense. Added that sentence. </p>

JamesAllingham commented on 0_intro_to_jax_for_ml.ipynb:
```diff
@@ -0,0 +1,2081 @@
+{
```
<p>Also opened by default for me.</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/deep-learning-indaba/indaba-pracs-2022/pull/4/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='24'/>

KaleabTessera replied:
<p>Made hidden. </p>

JamesAllingham commented on 0_intro_to_jax_for_ml.ipynb:
```diff
@@ -0,0 +1,2081 @@
+{
```
<pre class="ql-syntax" spellcheck="false">Line #2.    key = jax.random.PRNGKey(42)
</pre><p>I'd split this key into two, to start showing some good habits for key management:</p><pre class="ql-syntax" spellcheck="false">key1, key2 = jax.random.split(jax.random.PRNGKey(42), num=2)
x = jax.random.normal(key1, shape=(size, size))
y = jax.random.normal(key2, shape=(size, size))
</pre>
<br/>

 _Reply via <a href='https://app.reviewnb.com/deep-learning-indaba/indaba-pracs-2022/pull/4/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='40'/>

KaleabTessera replied:
<p>Agreed, will fix this. </p>

JamesAllingham commented on 0_intro_to_jax_for_ml.ipynb:
```diff
@@ -0,0 +1,2081 @@
+{
```
<p>Nit: <code>jit</code>,<code>grad</code>,<code>vmap</code>,<code>pmap</code>.</p><p> </p><p>Maybe mention examples of side effects? E.g., mutating state, printing, etc. And perhaps introduce the idea of a 'pure function' i.e., one without side effects, since this is a useful term when working with jax?</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/deep-learning-indaba/indaba-pracs-2022/pull/4/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='45'/>

sbodenstein replied:
<p>+1 to introduce pure functions. This is really the magic that allows for things like vmap etc.</p>

KaleabTessera replied:
@JamesAllingham @sbodenstein  By popular demand, I have added a section on pure functions (https://colab.research.google.com/github/deep-learning-indaba/indaba-pracs-2022/blob/prac1-intro-ml-using-jax/0_intro_to_ml_using_jax.ipynb#scrollTo=fT56qxXzTVKZ)! :fire: Please have a look and thanks for the suggestions!

JamesAllingham commented on 0_intro_to_jax_for_ml.ipynb:
```diff
@@ -0,0 +1,2081 @@
+{
```
<p>Nit: <code>jit</code> and <code>grad</code> rather than <code>JIT</code> and <code>GRAD</code> .    </p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/deep-learning-indaba/indaba-pracs-2022/pull/4/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='46'/>

KaleabTessera replied:
<p>Done. </p>

JamesAllingham commented on 0_intro_to_jax_for_ml.ipynb:
```diff
@@ -0,0 +1,2081 @@
+{
```
<p>Opened by default for me.</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/deep-learning-indaba/indaba-pracs-2022/pull/4/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='52'/>

KaleabTessera replied:
<p>Fixed. </p>

JamesAllingham commented on 0_intro_to_jax_for_ml.ipynb:
```diff
@@ -0,0 +1,2081 @@
+{
```
<p>Opened by default for me.</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/deep-learning-indaba/indaba-pracs-2022/pull/4/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='53'/>

KaleabTessera replied:
<p>Fixed, thanks!</p>

JamesAllingham commented on 0_intro_to_jax_for_ml.ipynb:
```diff
@@ -0,0 +1,2081 @@
+{
```
<p>I think it might be worth briefly explaining what caching is. E.g., "... compile and cache, that is save the compiled version so that it doesn't need to be recompiled everytime we call it, a version ...". Perhaps also linking to Wikipedia (or perhaps some better resource) to give a longer description of caching. I just worry that many folks won't know the term.</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/deep-learning-indaba/indaba-pracs-2022/pull/4/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='56'/>

KaleabTessera replied:
<p>True. Added the decription and link to the first <code>jit</code> section.  </p>

JamesAllingham commented on 0_intro_to_jax_for_ml.ipynb:
```diff
@@ -0,0 +1,2081 @@
+{
```
<pre class="ql-syntax" spellcheck="false">Line #17.    data = {'JAX (no jit)':jax_avg_time, 'JAX (with jit)':jax_jit_avg_time}
</pre><p>Nit: add spaces after ':'.</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/deep-learning-indaba/indaba-pracs-2022/pull/4/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='63'/>

sbodenstein replied:
<p>Nit: Jax -&gt; JAX</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/deep-learning-indaba/indaba-pracs-2022/pull/4/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='6'/>

JamesAllingham commented on 0_intro_to_jax_for_ml.ipynb:
```diff
@@ -0,0 +1,2081 @@
+{
```
<p>I'm a little confused here.  $\Delta f(x)$ should be $\nabla f(x)$ right? and <code>grad(f(x))</code> should be <code>grad(f)(x)</code>? </p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/deep-learning-indaba/indaba-pracs-2022/pull/4/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='64'/>

KaleabTessera replied:
<p>Thanks for this! Def a bug of late night math. This is fixed now :) </p>

KaleabTessera replied:
<p>Also switched to Lagrange's notation for simplicity. </p>

JamesAllingham commented on 0_intro_to_jax_for_ml.ipynb:
```diff
@@ -0,0 +1,2081 @@
+{
```
<p>Maybe print these numbers, otherwise it feels like running the cell doesn't do anything useful :P </p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/deep-learning-indaba/indaba-pracs-2022/pull/4/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='68'/>

KaleabTessera replied:
<p>I don't want to print these values since they are the answers to next question . </p>

JamesAllingham commented on 0_intro_to_jax_for_ml.ipynb:
```diff
@@ -0,0 +1,2081 @@
+{
```
<p>Should be hidden by default to avoid spoiling the answer.</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/deep-learning-indaba/indaba-pracs-2022/pull/4/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='70'/>

KaleabTessera replied:
<p>Will do. </p>

JamesAllingham commented on 0_intro_to_jax_for_ml.ipynb:
```diff
@@ -0,0 +1,2081 @@
+{
```
<p>I'd also print this.</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/deep-learning-indaba/indaba-pracs-2022/pull/4/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='72'/>

KaleabTessera replied:
<p>Same as above, don't want to print this since it is an answer to the next question.</p>

JamesAllingham commented on 0_intro_to_jax_for_ml.ipynb:
```diff
@@ -0,0 +1,2081 @@
+{
```
<p>Nit:<code>VMAP</code> -&gt;  <code>vmap</code> and <code>PMAP</code> -&gt;  <code>pmap</code></p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/deep-learning-indaba/indaba-pracs-2022/pull/4/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='78'/>

KaleabTessera replied:
<p>Done. </p>

JamesAllingham commented on 0_intro_to_jax_for_ml.ipynb:
```diff
@@ -0,0 +1,2081 @@
+{
```
<p>Nit: <code>VMAP</code> -&gt; <code>vmap</code>   </p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/deep-learning-indaba/indaba-pracs-2022/pull/4/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='79'/>

KaleabTessera replied:
<p>Done. </p>

JamesAllingham commented on 0_intro_to_jax_for_ml.ipynb:
```diff
@@ -0,0 +1,2081 @@
+{
```
<p>To be as clear as possible I'd do:</p><pre class="ql-syntax" spellcheck="false">batch_size = 3
batched_x = np.arange(15).reshape((batch_size, -1))
</pre>
<br/>

 _Reply via <a href='https://app.reviewnb.com/deep-learning-indaba/indaba-pracs-2022/pull/4/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='86'/>

KaleabTessera replied:
<p>Good point, done. </p>

JamesAllingham commented on 0_intro_to_jax_for_ml.ipynb:
```diff
@@ -0,0 +1,2081 @@
+{
```
<p>You seem to be giving the cprrect answer here by default?</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/deep-learning-indaba/indaba-pracs-2022/pull/4/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='89'/>

KaleabTessera replied:
<p>This has changed now. </p>

JamesAllingham commented on 0_intro_to_jax_for_ml.ipynb:
```diff
@@ -0,0 +1,2081 @@
+{
```
<p>I'd avoid the list comprehension here (since some folks might not be familiar with them) and rather manually construct the list using a good old fashioned for loop.</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/deep-learning-indaba/indaba-pracs-2022/pull/4/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='91'/>

KaleabTessera replied:
<p>Updated. </p>

JamesAllingham commented on 0_intro_to_jax_for_ml.ipynb:
```diff
@@ -0,0 +1,2081 @@
+{
```
<pre class="ql-syntax" spellcheck="false">Line #6.    # Run it on a single vecor
</pre><p>I think this single vector example might actually cause more confusion than help. Since I don't think this issue will actually come up. Not sure though...</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/deep-learning-indaba/indaba-pracs-2022/pull/4/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='95'/>

KaleabTessera replied:
<p>Valid. Let's see what other tutors say.</p>

JamesAllingham commented on 0_intro_to_jax_for_ml.ipynb:
```diff
@@ -0,0 +1,2081 @@
+{
```
<p>Spelling :)</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/deep-learning-indaba/indaba-pracs-2022/pull/4/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='96'/>

KaleabTessera replied:
<p>Fixed. Thanks! </p>

JamesAllingham commented on 0_intro_to_jax_for_ml.ipynb:
```diff
@@ -0,0 +1,2081 @@
+{
```
<p>I'd hide by default.</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/deep-learning-indaba/indaba-pracs-2022/pull/4/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='101'/>

KaleabTessera replied:
<p>Done. </p>

JamesAllingham commented on 0_intro_to_jax_for_ml.ipynb:
```diff
@@ -0,0 +1,2081 @@
+{
```
<p>Nit: missing ')', 'has' -&gt; 'have'</p><p> </p><p>I am not sure that I would use <code>pmap</code> to create these matrices. I think it might be more clear to reuse the previous matrices and focus on explaining <code>pmap</code> for the batched dot product?  </p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/deep-learning-indaba/indaba-pracs-2022/pull/4/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='107'/>

KaleabTessera replied:
<p>Made spelling fixes. So using pmap makes it a <code>ShardedDeviceArray</code> , which is what you need to <code>pmap</code> effectively.   </p>

JamesAllingham commented on 0_intro_to_jax_for_ml.ipynb:
```diff
@@ -0,0 +1,2081 @@
+{
```
<p>Might be worth mentioning that the speedup isn't that impressive in this particular case, but that for some problems you can get an almost Nx speedup if you have N devices?</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/deep-learning-indaba/indaba-pracs-2022/pull/4/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='115'/>

KaleabTessera replied:
<p>Yep. Added. </p>

JamesAllingham commented:
Looks good so far! Keen to see what it looks like with part 2.

Took me about 45 mins to go through, but that includes the time spent typing out comments. It is hard for me to judge how long this would take for a jax newbie, but I think if they skipped the `pmap` section (which we should encourage for beginners) you'd be looking at around 1hr which seems about right?

Like Javi, I was unable to access the quiz. 

I agree with Javi's other points too. However, w.r.t 
> I would uncollapse cells by default because otherwise it is possible to get confused and skip a section

I would just say that that the code cells with titles should by default be hidden, to remove extra complexity where it isn't necessary. And same goes for the solution cells :P

ruanvdmerwe commented on 0_intro_to_jax_for_ml.ipynb:
```diff
@@ -0,0 +1,2081 @@
+{
```
<p>A small thing that I could perhaps see happening is that step 2 is interpreted as running both cells below, which can mess up the rest of the notebook. </p><p> </p><p>I might move this block after defining <code>parameters_list</code>, or explicitly state that they should only run the cell where  <code>parameters_list</code> is defined once. </p><p> </p><p> </p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/deep-learning-indaba/indaba-pracs-2022/pull/4/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='126'/>

KaleabTessera replied:
<p>Thanks so much ! This will def be fixed! </p>

sbodenstein commented on 0_intro_to_jax_for_ml.ipynb:
```diff
@@ -0,0 +1,2081 @@
+{
```
<p>One general comment: it can be very useful for people coming from other frameworks (most notably PyTorch) to get some idea of what problem JAX is trying to solve. Why was JAX created in a world that already has PyTorch, and why should they be interested in learning about it in these tutorials? I can imagine this coming up a lot. (this could go at the end of the tutorial for example). </p><p> </p><blockquote>JAX is <strong>not</strong> a replacement for Pytorch or Tensorflow, but a lower-level library commonly used with higher-level neural network libraries such as <a href="https://github.com/deepmind/dm-haiku" target="_blank">Haiku</a> or <a href="https://github.com/google/flax" target="_blank">Flax</a>.</blockquote><p>This is a confusing statement. JAX is 100% intended to be a replacement for PyTorch and TensorFlow. It can be used in its plain form, but it is often used via libraries that make lots of things simpler (precisely as TensorFlow was always used via Keras, TensorFlow Slim, etc etc).</p><p> </p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/deep-learning-indaba/indaba-pracs-2022/pull/4/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='0'/>

KaleabTessera replied:
@sbodenstein Thanks for comments. 

- On "why JAX" - I guess I was trying to answer this in the https://colab.research.google.com/github/deep-learning-indaba/indaba-pracs-2022/blob/prac1-intro-ml-using-jax/0_intro_to_ml_using_jax.ipynb#scrollTo=TSj972IWxTo2 section (JAX is backed agnostic and all the useful transforms), but I get what you are saying. Do you think having this in the intro or conclusion would work best? Like a "Why JAX" section. 

- On the "JAX is not a replacement ... " part, I have updated this (https://colab.research.google.com/github/deep-learning-indaba/indaba-pracs-2022/blob/prac1-intro-ml-using-jax/0_intro_to_ml_using_jax.ipynb#scrollTo=Enx0WUr8tIPf) . Please let me know what you think :)

sbodenstein commented on 0_intro_to_jax_for_ml.ipynb:
```diff
@@ -0,0 +1,2081 @@
+{
```
<p>Nit: prng -&gt; PRNG</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/deep-learning-indaba/indaba-pracs-2022/pull/4/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='25'/>

KaleabTessera replied:
<p>Updated. </p>

sbodenstein commented on 0_intro_to_jax_for_ml.ipynb:
```diff
@@ -0,0 +1,2081 @@
+{
```
<p>Nit: 'is more explicit' to 'is explicit'. More explicit implies that parts of it are not explicit. But it is 100% explicit in JAX.</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/deep-learning-indaba/indaba-pracs-2022/pull/4/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='28'/>

KaleabTessera replied:
<p>Updated. </p>

sbodenstein commented on 0_intro_to_jax_for_ml.ipynb:
```diff
@@ -0,0 +1,2081 @@
+{
```
<p>Nit: XLA has nothing to do with <code>vmap</code>.It can run <code>vmap</code> without XLA.   </p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/deep-learning-indaba/indaba-pracs-2022/pull/4/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='34'/>

KaleabTessera replied:
<p>Fair point, I have updated this. </p>

sbodenstein commented on 0_intro_to_jax_for_ml.ipynb:
```diff
@@ -0,0 +1,2081 @@
+{
```
<p>Nit: gpu/tpu -&gt; GPU/TPU</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/deep-learning-indaba/indaba-pracs-2022/pull/4/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='35'/>

KaleabTessera replied:
<p>Done. </p>

sbodenstein commented on 0_intro_to_jax_for_ml.ipynb:
```diff
@@ -0,0 +1,2081 @@
+{
```
<p>The key to getting things like vmap is that jax functions are pure. Both of these JAX design decisions only make sense in that light (you can't have mutable rng state in pure functions, and you can't have jax pure functions mutating input arrays. </p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/deep-learning-indaba/indaba-pracs-2022/pull/4/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='15'/>

KaleabTessera replied:
Added pure functions section(https://colab.research.google.com/github/deep-learning-indaba/indaba-pracs-2022/blob/prac1-intro-ml-using-jax/0_intro_to_ml_using_jax.ipynb#scrollTo=fT56qxXzTVKZ).

KaleabTessera commented:
> I did the intro prac part 1.
> 
> Timing:
> 
> It took me 42 minutes, although I got distracted reading through documentation and with slack messages various times (Perhaps this simulates the lively in-person indaba environment.) Without stopping, I think it would have taken 30 minutes. If I had not read all the cells and understood all of the code (since I could guess what it was going to do from experience with jax / numpy) I think it would have taken 15 minutes.
> 
> Comments / potential improvements:
> 
> * I would uncollapse cells by default because otherwise it is possible to get confused and skip a section
> * In the random key explanation for jax, it is not completely obvious that in `key, subkey = random.split(key)` , the returned key is different from the inputted key (since both have the same name). This could cause some confusion.
> * In the ReLU example, I would make the provided solution use` jnp.where`  since the `jnp.maximum`  solution is trivial. This way people will probably use `jnp.maximum` in their solution but can learn something from the solved exercise as well.
> * The `min_max(batched_x)` example has the solution already coded up and inputed by default. This is a bug I think.
> * For the parallelisation of the `min_max(batched_x)` section, the matrices used are huge. I let this run for 7 minutes on CPU and stopped it. I would make the leading dimension 500 instead of 5000 since otherwise everyone will need GPU allocated. I could not get GPU or TPU today for some reason. There are 2 or 3 cells where this is a problem.
> * The quiz form would not load in the notebook for me and I had to click the link to go to the quiz page explicitly

Thanks so much @JavierAntoran !!! 

Here is my feedback/updates

> I would uncollapse cells by default because otherwise it is possible to get confused and skip a section

- Done. 

> In the random key explanation for jax, it is not completely obvious that in key, subkey = random.split(key) , the returned key is different from the inputted key (since both have the same name). This could cause some confusion.

- Updated. Call it new_key. 

> In the ReLU example, I would make the provided solution use jnp.where since the jnp.maximum solution is trivial. This way people will probably use jnp.maximum in their solution but can learn something from the solved exercise as well.

- Updated to use jnp.where. 

> The min_max(batched_x) example has the solution already coded up and inputed by default. This is a bug I think.

**- Can you check exercise 1.7. Is this fine?** 

> For the parallelisation of the min_max(batched_x) section, the matrices used are huge. I let this run for 7 minutes on CPU and stopped it. I would make the leading dimension 500 instead of 5000 since otherwise everyone will need GPU allocated. I could not get GPU or TPU today for some reason. There are 2 or 3 cells where this is a problem.

- Made matrices much smaller. 

> The quiz form would not load in the notebook for me and I had to click the link to go to the quiz page explicitly

- Quiz for now is WIP.

lollcat commented on 0_intro_to_ml_using_jax.ipynb:
```diff
@@ -0,0 +1,3472 @@
+{
```
<p>Nitpick:</p><blockquote>an upcoming machine learning framework</blockquote><p>Could instead be</p><blockquote>a new machine learning framework</blockquote><p>given that jax is already quite widely used it doesn't feel "upcoming" to me (but I suppose it is still V0). </p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/deep-learning-indaba/indaba-pracs-2022/pull/4/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='0'/>

joaogui1 replied:
<p>Strong agree here!</p>

KaleabTessera replied:
<p>Updated. Thanks for feedback! </p>

lollcat commented on 0_intro_to_ml_using_jax.ipynb:
```diff
@@ -0,0 +1,3472 @@
+{
```
<p>"and oftentimes," -&gt; "and often"</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/deep-learning-indaba/indaba-pracs-2022/pull/4/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='7'/>

KaleabTessera replied:
<p>Done. </p>

lollcat commented on 0_intro_to_ml_using_jax.ipynb:
```diff
@@ -0,0 +1,3472 @@
+{
```
<p>Could the initialisation be made worse? (This lets the user hone in more on what a good fit is). Could the increments be made smaller for the slider? (e.g. see code below). Also could the plot be a bit more zoomed in so that the trend is more clear? </p><pre class="ql-syntax" spellcheck="false">b = -2.0 #@param {type:"slider", min:-5, max:5, step:0.5}
w = -0.02 #@param {type:"slider", min:-0.05, max:0.05, step:0.005}
print("Plotting line", w, "* x +", b)
parameters = [b, w]
parameters_list.append(parameters)
plot_basic_data(parameters_list,
&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; title="Observed data and my first predictions",axis_pad=10)
</pre>
<br/>

 _Reply via <a href='https://app.reviewnb.com/deep-learning-indaba/indaba-pracs-2022/pull/4/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='135'/>

KaleabTessera replied:
<p>Thanks! I have updated the init. I made the axis_pad=12 (for some values of w, when using axis_pad=10, it goes off the plot). </p>

JavierAntoran commented:
I went through the linear regression section and the Optax / Haiku section. In general I found the sections very clear and well put together so I just have some nits:

* In the `Formalizing the Loss Function` section
    * Residual lines should not be purple because they are had to distinguish from the red regression line. I did not realise they were purple until it said so in words in the next cell.

* In the `Why learn gradient descent then?` section: I would change the order of the `computational complexity and `more general` arguments. 

It is true that gradient descent can be cheaper than exact solutions. However, I would say the main reason people use gradient descent because is *analytical complexity* of exact solutions, not computational complexity. For the vast majority of settings, the optima can not be found in closed form. Thus, I think the second point `more general` should come first. 
For the  `computational complexity` point, I would add the clause “Even when a closed form solution is available...”. 

* Right before exercise 2.6:
	* “The fact that our model has very loss training loss “ -> The fact that our model has very **low** training loss 
	* The order 10 polynomial model makes massive predictions and thus the plot yscale becomes large, making it hard to see what is happening. I would set a fixed ylim
	* Something that is not addressed but is also a concern is overfitting the noise in the interpolation regime. Perhaps there is not enough time though…

* The end of the prac is somewhat anticlimactic. It could be nice to plot the optax optimisation trajectory as a visual payoff. We will probably see that it is somewhat different than the manually coded one due to Optax using the Adam optimiser instead of SGD.

JavierAntoran reviewed (dismissed):
I added some minor suggestions. Looks very polished overall!

joaogui1 commented on 0_intro_to_ml_using_jax.ipynb:
```diff
@@ -0,0 +1,3758 @@
+{
```
<p>I think you should add a small sentence saying that XLA is a compiler created by Google that allows one to target different kinds of accelerators (CPU, GPU and TPU)</p><p>I would also add that the transformations are in general composable and that JAX exposes a numpy API + the transformations</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/deep-learning-indaba/indaba-pracs-2022/pull/4/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='RESOLVED' data-cellIndex='7'/>

KaleabTessera replied:
<p>Thanks. I have updated this. </p>

joaogui1 commented on 0_intro_to_ml_using_jax.ipynb:
```diff
@@ -0,0 +1,3758 @@
+{
```
<pre class="ql-syntax" spellcheck="false">Line #3.    x = jnp.linspace(-np.pi,np.pi,100)
</pre><p>Nitpick: Add spacing between the arguments </p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/deep-learning-indaba/indaba-pracs-2022/pull/4/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='RESOLVED' data-cellIndex='17'/>

KaleabTessera replied:
<p>This is now fixed since everything is linted. </p>

joaogui1 commented on 0_intro_to_ml_using_jax.ipynb:
```diff
@@ -0,0 +1,3758 @@
+{
```
<p>I think you should mention how the clunky design is on purpose as you in general shouldn't do in-place mutation when using JAX</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/deep-learning-indaba/indaba-pracs-2022/pull/4/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='RESOLVED' data-cellIndex='24'/>

joaogui1 replied:
<p>I think you could add more explicitly that this means that while JAX implements the numpy API in general it means there's no <code>jax.numpy.random</code></p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/deep-learning-indaba/indaba-pracs-2022/pull/4/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='27'/>

joaogui1 replied:
<p>There's now a <code>jax.debug.print</code> <a href="https://jax.readthedocs.io/en/latest/debugging/print_breakpoint.html" target="_blank">function</a>!</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/deep-learning-indaba/indaba-pracs-2022/pull/4/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='RESOLVED' data-cellIndex='95'/>

ElanVB replied:
<p>I revieved an error (jax or jnp not defined or something like that) after swapping to a TPU and running this cell. So, I hit "run all" to make sure everything was imported properly and then my runtime unexpectedly crashed (repeatedly) when using jax to plot sin(x). I think the <code>jax.tools.colab_tpu.setup_tpu()</code> call must not be working properly because I seem to be getting CUDA errors that make my Kernal restart:</p><p><code>external/org_tensorflow/tensorflow/stream_executor/cuda/cuda_driver.cc:265] failed call to cuInit: CUDA_ERROR_NO_DEVICE: no CUDA-capable device is detected</code></p><p>Also, if we are swapping runtimes here, we might need to let them know they need to re-run the first few cells (or just put that code here), so that libraries are imported.</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/deep-learning-indaba/indaba-pracs-2022/pull/4/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='126'/>

joaogui1 commented on 0_intro_to_ml_using_jax.ipynb:
```diff
@@ -0,0 +1,3758 @@
+{
```
<p>How about also mentioning <code>has_aux</code>?</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/deep-learning-indaba/indaba-pracs-2022/pull/4/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='RESOLVED' data-cellIndex='82'/>

KaleabTessera replied:
<p>Added paragraph on this, thanks! </p>

joaogui1 commented on 0_intro_to_ml_using_jax.ipynb:
```diff
@@ -0,0 +1,3758 @@
+{
```
<p>This also mean that it's ok to use global <strong>constants </strong>inside jax functions</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/deep-learning-indaba/indaba-pracs-2022/pull/4/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='RESOLVED' data-cellIndex='101'/>

KaleabTessera replied:
<p>Nice point. Will add a sentence on this. </p>

ElanVB commented on 0_intro_to_ml_using_jax.ipynb:
```diff
@@ -0,0 +1,3758 @@
+{
```
<pre class="ql-syntax" spellcheck="false">Line #2.    x = jnp.linspace(-np.pi,np.pi,100)
</pre><p>Very small things, might be better to change the "np.pi"s to "jnp.pi"s. Same with future uses. Just so that we show that it's best to use jnp instead of np as far as possible.</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/deep-learning-indaba/indaba-pracs-2022/pull/4/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='RESOLVED' data-cellIndex='14'/>

KaleabTessera replied:
<p>Will do, thanks! </p>

ElanVB commented on 0_intro_to_ml_using_jax.ipynb:
```diff
@@ -0,0 +1,3758 @@
+{
```
<pre class="ql-syntax" spellcheck="false">Line #6.    new_key, subkey = random.split(key)
</pre><p>I haven't used JAX before, so let. me know if this is a naming convention. It might be better to give slightly more intuitive names for the keys. Something like "new_split_key" (one we will continue to split) and "new_generator_key" (one we will use to generate a new random number).</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/deep-learning-indaba/indaba-pracs-2022/pull/4/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='RESOLVED' data-cellIndex='38'/>

KaleabTessera replied:
<p>Yeah, I think this more the naming covention thing. Usually, people just say <code>key, subkey = ...</code>.  I added <code>new_key</code> to be more explicit.   </p>

ElanVB commented on 0_intro_to_ml_using_jax.ipynb:
```diff
@@ -0,0 +1,3758 @@
+{
```
<p>This link seems to be broken. I can't seem to find a working link for those design notes.</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/deep-learning-indaba/indaba-pracs-2022/pull/4/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='RESOLVED' data-cellIndex='39'/>

KaleabTessera replied:
<p>Updated with a new link, thanks! </p>

ElanVB commented on 0_intro_to_ml_using_jax.ipynb:
```diff
@@ -0,0 +1,3758 @@
+{
```
<pre class="ql-syntax" spellcheck="false">Line #1.    f = lambda x: 6*x**4 - 9*x + 4
</pre><p>Why do we use a lambda function here instead of a regular function definition for f?  </p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/deep-learning-indaba/indaba-pracs-2022/pull/4/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='71'/>

KaleabTessera replied:
<p>I guess that because it is close to the mathematical function f. </p>

ElanVB commented on 0_intro_to_ml_using_jax.ipynb:
```diff
@@ -0,0 +1,3758 @@
+{
```
<p>Just. tocheck my understanding so that I can answer questions related to this:</p><p>JAX: can get gradient for any python function (may be some limitations, not sure if it needs to be pure, etc)</p><p>TF and PyTorch: we have to use the frameworks functions (torch.relu, etc) if we want it to be able to calculate the grad.</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/deep-learning-indaba/indaba-pracs-2022/pull/4/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='RESOLVED' data-cellIndex='85'/>

KaleabTessera replied:
<p>So TF and PyTorch would have their own autograd things, but jax is less cleaner (e.g. no <code>.backward()</code> in PyTorch or <code>tape.gradient</code> in tf). You can just define a function naturally and get the grad, I also think it can diff over more diverse things like loops, branches, recursion, and closures.   </p>

ElanVB commented on 0_intro_to_ml_using_jax.ipynb:
```diff
@@ -0,0 +1,3758 @@
+{
```
<p>Should we explain why this is? (I'm not sure but I would guess that JAX saw that the output of the impure_print_side_effect function does not change based on the return of the print function)</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/deep-learning-indaba/indaba-pracs-2022/pull/4/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='RESOLVED' data-cellIndex='92'/>

KaleabTessera replied:
<p>Yes, your reason is right. We do mention in the next cell that changing shape triggers a retrace. Do you think that is explicit enough?  </p>

ElanVB commented on 0_intro_to_ml_using_jax.ipynb:
```diff
@@ -0,0 +1,3758 @@
+{
```
<p>Slightly further on we say "So&nbsp;<code>vmap</code>&nbsp;should be similar in performance to manually vectorized code (if everything is implemented well), and much better than naively vectorized code (i.e. for loops)". Should we rather use the world "manually" here and "naively" above? I found myself wondering if "natively" and "naively" were supposed to be the same word and there was a typo.</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/deep-learning-indaba/indaba-pracs-2022/pull/4/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='RESOLVED' data-cellIndex='118'/>

KaleabTessera replied:
<p>Yes, great point! Changed! </p>

ElanVB commented on 0_intro_to_ml_using_jax.ipynb:
```diff
@@ -0,0 +1,3758 @@
+{
```
<pre class="ql-syntax" spellcheck="false">Line #13.        assert jnp.equal(error, correct_error[i]), f"Incorrect implementation. Value: {error} Expected Value: {correct_error[i]}. Parameters (b,w,x_i,y_i): {b,w,x_i,y_i} "
</pre><p>We might need to put some precision tolarance in here. I got the following error:</p><p><code>Incorrect implementation. Value: 105.4729 Expected Value: 105.47291. Parameters (b,w,x_i,y_i): (3.77, 0.05, 210, 4)</code></p><p> </p><p>Oh, I see why my implementations precision is different. I used base python instead of jnp (we should maybe explicitly say "use jnp to implement...")</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/deep-learning-indaba/indaba-pracs-2022/pull/4/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='RESOLVED' data-cellIndex='166'/>

KaleabTessera replied:
<p>I think we do mention that jnp.dot is a useful function. Do you think that is enough?</p>

ElanVB commented on 0_intro_to_ml_using_jax.ipynb:
```diff
@@ -0,0 +1,3758 @@
+{
```
<pre class="ql-syntax" spellcheck="false">Line #7.      return loss / 2 * len(x_data_list)
</pre><p>I think there is a small error here. Shouldn't it be <code>loss / (2 * len(x_data_list))</code> ? What we have is <code>(loss x N) / 2</code>  rather than <code>loss / (2N)</code> . </p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/deep-learning-indaba/indaba-pracs-2022/pull/4/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='170'/>

KaleabTessera replied:
<p>You are right, this was a bug. Thank you! </p>

ElanVB commented on 0_intro_to_ml_using_jax.ipynb:
```diff
@@ -0,0 +1,3758 @@
+{
```
<p>It might be worth giving a little input about what <code>argnums=(0, 1)</code> is doing. (Telling JAX to calculate the partial gradients with respect to the first two parameters of <code>loss_function</code>, right?)</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/deep-learning-indaba/indaba-pracs-2022/pull/4/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='RESOLVED' data-cellIndex='184'/>

KaleabTessera replied:
<p>Yep, added a description. Thnaks! </p>

ElanVB commented on 0_intro_to_ml_using_jax.ipynb:
```diff
@@ -0,0 +1,3758 @@
+{
```
<p>Shall we all come up with some possible answers to these questions so that we don't have to think about it during the prac?</p><p> </p><ol><li>Model complexity (too high or too low resulting in over- or under-fitting), Poor optimisation (too high or low learning rate, noisy optimisation landscape, etc)</li><li>Collect more data, make sure your data covers the relevant domain well, add some regularisation techniques to your optimisation process (Weight decay, dropout, etc)</li><li>we assume the seen and unseed data are distributed similarly (look roughly the same): they cover the same domain, data spread over the domain is similar, the data is spread over the domain in the same way</li></ol>
<br/>

 _Reply via <a href='https://app.reviewnb.com/deep-learning-indaba/indaba-pracs-2022/pull/4/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='201'/>

KaleabTessera replied:
<p>Let's add this the prac channel so that we have these great discussion points! </p>

KaleabTessera commented:
@ElanVB @joaogui1 I will merge the current version of this. I will address your comments in future PRs (we need a version of this prac in main for now) :)

JavierAntoran reviewed (approved):
LGTM!

ruanvdmerwe commented on 0_intro_to_ml_using_jax.ipynb:
```diff
@@ -0,0 +1,3784 @@
+{
```
<p>Very good questions here that can cause GOOD rabbit holes. </p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/deep-learning-indaba/indaba-pracs-2022/pull/4/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='RESOLVED' data-cellIndex='199'/>

ruanvdmerwe commented on 0_intro_to_ml_using_jax.ipynb:
```diff
@@ -0,0 +1,3784 @@
+{
```
<p>Maybe just show here how it would have been called if it is something stochastic. </p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/deep-learning-indaba/indaba-pracs-2022/pull/4/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='209'/>

KaleabTessera replied:
<p>I think time is an issue. </p>

ruanvdmerwe commented on 0_intro_to_ml_using_jax.ipynb:
```diff
@@ -0,0 +1,3784 @@
+{
```
<pre class="ql-syntax" spellcheck="false">Line #38.        train_dataset = train_dataset.shuffle(10 * train_batch_size, seed=seed)
</pre><p>I might bring attention here that shuffling happens before batching when creating tensorflow dataset. Such a typical pitfall.</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/deep-learning-indaba/indaba-pracs-2022/pull/4/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='225'/>

KaleabTessera replied:
<p>Added link to https://www.tensorflow.org/guide/data_performance in the comments :) </p>

ruanvdmerwe commented on 0_intro_to_ml_using_jax.ipynb:
```diff
@@ -0,0 +1,3784 @@
+{
```
<pre class="ql-syntax" spellcheck="false">Line #78.    # Training &amp; evaluation loop.
</pre><p>Could maybe add in livelossplot here as most of the other practicals also use it now, very NIT like all my current comments</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/deep-learning-indaba/indaba-pracs-2022/pull/4/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='227'/>

KaleabTessera replied:
<p>Will add, thanks! </p>

ruanvdmerwe reviewed (commented):
I have very NIT comments and I think the latest comments added are super valuable. But in terms of having a working practical that is really good and can be seen, I think this meets the criteria.
